librz/analysis/rzil/rzil_trace.c

```diff
@@ -0,0 +1,220 @@
+// SPDX-FileCopyrightText: 2021 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_analysis.h>
+
+/**
+ * RZIL trace should also these info
+ * 1. mem.read address & data
+ * 2. mem.write address & data
+ * 3. reg.read name & data
+ * 4. reg.write name & data
+ **/
+
+static void htup_vector_free(HtUPKv *kv) {
+	rz_vector_free(kv->value);
+}
+
+RZ_API RzAnalysisRzilTrace *rz_analysis_rzil_trace_new(RzAnalysis *analysis, RzAnalysisRzil *rzil) {
+	rz_return_val_if_fail(rzil, NULL);
+	size_t i;
+	RzAnalysisEsilTrace *trace = RZ_NEW0(RzAnalysisEsilTrace);
+	if (!trace) {
+		return NULL;
+	}
+
+	// TODO : maybe we could remove memory && register in rzil trace ?
+	trace->registers = ht_up_new(NULL, htup_vector_free, NULL);
+	if (!trace->registers) {
+		goto error;
+	}
+	trace->memory = ht_up_new(NULL, htup_vector_free, NULL);
+	if (!trace->memory) {
+		goto error;
+	}
+	trace->instructions = rz_pvector_new((RzPVectorFree)rz_analysis_il_trace_instruction_free);
+	if (!trace->instructions) {
+		goto error;
+	}
+
+	// TODO : Integrate with stack panel in the future
+
+	// Save initial registers arenas
+	for (i = 0; i < RZ_REG_TYPE_LAST; i++) {
+		RzRegArena *a = analysis->reg->regset[i].arena;
+		RzRegArena *b = rz_reg_arena_new(a->size);
+		if (!b) {
+			goto error;
+		}
+		if (b->bytes && a->bytes && b->size > 0) {
+			memcpy(b->bytes, a->bytes, b->size);
+		}
+		trace->arena[i] = b;
+	}
+	return trace;
+error:
+	eprintf("Fail to init RZIL trace\n");
+	rz_analysis_esil_trace_free(trace);
+	return NULL;
+}
+
+RZ_API void rz_analysis_rzil_trace_free(RzAnalysisEsilTrace *trace) {
+	size_t i;
+	if (trace) {
+		ht_up_free(trace->registers);
+		ht_up_free(trace->memory);
+		for (i = 0; i < RZ_REG_TYPE_LAST; i++) {
+			rz_reg_arena_free(trace->arena[i]);
+		}
+		rz_pvector_free(trace->instructions);
+		trace->instructions = NULL;
+		RZ_FREE(trace);
+	}
+}
+
+// IL trace wrapper of ESIL
+static inline void rzil_add_mem_trace(RzAnalysisRzilTrace *rtrace, RzILTraceMemOp *mem) {
+	RzILTraceInstruction *instr_trace = rz_analysis_esil_get_instruction_trace(rtrace, rtrace->idx);
+	rz_analysis_il_trace_add_mem(instr_trace, mem);
+}
+
+static inline void rzil_add_reg_trace(RzAnalysisRzilTrace *rtrace, RzILTraceRegOp *reg) {
+	RzILTraceInstruction *instr_trace = rz_analysis_esil_get_instruction_trace(rtrace, rtrace->idx);
+	rz_analysis_il_trace_add_reg(instr_trace, reg);
+}
+
+// buf limit 32
+static void bv_to_databuf(ut8 *buf, RzILBitVector bv) {
+	// rz_il_bv_prepend_zero(bv, 128 - bv->len);
+	//	if (bv->_elem_len != 16) {
+	//		RZ_LOG_ERROR("BAD SIZE\n");
+	//		return;
+	//	}
+}
+
+static void rz_analysis_rzil_trace_focus_mem_read(RzAnalysis *analysis, RzAnalysisRzil *rzil, RzILOp single_op) {
```

Comment:
Add `RZ_NONNULL` for the `rzil` and `single_op`. Also please check the `rzil_op` within `rz_return_if_fail`.